Build silence operator for arm64

QuentinBisson opened the pull request:
Solving: https://github.com/giantswarm/silence-operator/issues/497

## Checklist

- [x] Update changelog in CHANGELOG.md.

QuentinBisson committed:
try out multi-arch

QuentinBisson commented on .circleci/config.yml:
```diff
@@ -22,20 +22,27 @@ workflows:
 
     - architect/go-build:
-        name: go-build
+        name: go-build-<<matrix.architectures>>
```
```suggestion
        name: go-build-<<matrix.architecture>>
```

QuentinBisson committed:
Update .circleci/config.yml

QuentinBisson committed:
Update config.yml

QuentinBisson committed:
Merge branch 'main' into try-multi-arch

TheoBrigitte commented on .circleci/config.yml:
```diff
@@ -22,20 +22,25 @@ workflows:
-        - go-build
+          - architect/go-build
+        image: giantswarm/silence-operator
+        platforms: "linux/amd64,linux/arm64"
```
Why isn't this a matrix ? How can we push both arch in the same job ?

QuentinBisson replied:
docker buildx builds the image and push it as the same time

QuentinBisson replied:
So it actually pushed to all registries with a one line command

TheoBrigitte replied:
Alright

TheoBrigitte commented on Dockerfile:
```diff
@@ -1,10 +1,13 @@
+# Use a build stage to select the correct binary based on TARGETPLATFORM
+FROM alpine AS binary-selector
+ARG TARGETPLATFORM
+COPY silence-operator-* /binaries/
+RUN arch="${TARGETPLATFORM#linux/}" && cp "/binaries/silence-operator-linux-${arch}" /silence-operator
+
 # Use distroless as minimal base image to package the silence-operator binary
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
 FROM gcr.io/distroless/static:nonroot
 WORKDIR /
-
-ADD silence-operator silence-operator
-
+COPY --from=binary-selector /silence-operator /silence-operator
```
This all sounds very strange and twisted to me, I would rather keep the Dockerfile simple as it was. This is a smell that we are not doing things right beforehand. I think the pipeline should be made simple where a binary is built and then added into the docker image. There should be no need to "select" the right binary here, that's very error prone.

Just copy a single binary into the image.

QuentinBisson replied:
You cannot select 1 binary because binaires are build for each platform but we can have a multi-arch image (so 1 shared image that work on all Platforms)

QuentinBisson replied:
I correct myself, we could without this but that means we would need to build the binary into the docker container to have 1 binary  per platform. This might bé where we are going but this is not how architect works today

TheoBrigitte commented:
Out of curiosity I gave this a try on different architecture, and it did not work on any Arm processor I tried, one being `aarch64` and the other `armv7`.

* Regular amd64 laptop :heavy_check_mark:  

```
$ uname -a
Linux x1 6.6.94-1-MANJARO #1 SMP PREEMPT_DYNAMIC Thu Jun 19 17:12:31 UTC 2025 x86_64 GNU/Linux
$ lscpu
Architecture:                x86_64
  CPU op-mode(s):            32-bit, 64-bit
  Address sizes:             39 bits physical, 48 bits virtual
  Byte Order:                Little Endian
CPU(s):                      8
  On-line CPU(s) list:       0-7
Vendor ID:                   GenuineIntel
  Model name:                11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
$ docker run --rm -it giantswarm/silence-operator:0.17.0-0795405a6ede65117c232f52f3cba58970d071a2 --help
Usage of /silence-operator:
...
```

* Raspberry Pi 3 Model B Plus Rev 1.3 :x: 

```
$ uname -a
Linux zul 6.1.21-v7+ #1642 SMP Mon Apr  3 17:20:52 BST 2023 armv7l GNU/Linux
$ lscpu
Architecture:           armv7l
  Byte Order:           Little Endian
CPU(s):                 4
  On-line CPU(s) list:  0-3
Vendor ID:              ARM
  Model name:           Cortex-A53
$ docker run --rm -it giantswarm/silence-operator:0.17.0-0795405a6ede65117c232f52f3cba58970d071a2 --help
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm/v7) and no specific platform was requested
exec /silence-operator: exec format error
```

* ARM Altra processor from Ampere :x: 

```
$ uname -a
Linux instance-20250104-2100 5.15.0-309.180.4.el9uek.aarch64 #2 SMP Wed May 21 07:07:17 PDT 2025 aarch64 aarch64 aarch64 GNU/Linux
$ lscpu
Architecture:                aarch64
  CPU op-mode(s):            32-bit, 64-bit
  Byte Order:                Little Endian
CPU(s):                      4
  On-line CPU(s) list:       0-3
Vendor ID:                   ARM
  BIOS Vendor ID:            QEMU
  Model name:                Neoverse-N1
$ docker run --rm -it giantswarm/silence-operator:0.17.0-0795405a6ede65117c232f52f3cba58970d071a2 --help
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
WARNING: image platform (linux/amd64) does not match the expected platform (linux/arm64)
{"msg":"exec container process `/silence-operator`: Exec format error","level":"error","time":"2025-07-03T17:08:47.841517Z"}
```

QuentinBisson commented:
I'll have to figure this out

TheoBrigitte commented:
I would also suggest adding steps in the workflow to ensure the built images are working with the different architectures. So I don't have to run those tests and all my machines every time :D

QuentinBisson committed:
Update Dockerfile

QuentinBisson commented on Dockerfile:
```diff
@@ -1,10 +1,12 @@
-
-ADD silence-operator silence-operator
-
+COPY --from=binary-selector /binaries/silence-operator-${TARGETPLATFORM#linux/} /silence-operator
```
```suggestion
COPY --from=binary-selector /binaries/silence-operator-linux-${TARGETPLATFORM#linux/} /silence-operator
```

QuentinBisson committed:
Update Dockerfile

QuentinBisson committed:
Merge branch 'main' into try-multi-arch

QuentinBisson committed:
validate images

QuentinBisson committed:
fix

QuentinBisson committed:
Build multi-arch image via architect orb 8.2.2

Bumps `giantswarm/architect` orb to 8.2.2 and switches to the
single-job `go-build` (`architectures: linux/amd64,linux/arm64`) +
`push-to-registries` (`multiarch: true`) path. Buildx produces the
manifest list directly, so the bespoke `validate-silence-operator-multiarch`
job and its `docker_image_tag.txt` workspace handoff are removed.

The Dockerfile drops the binary-selector stage and uses
`COPY silence-operator-linux-${TARGETARCH}` — buildx scopes the
build context per platform so the matching binary is the only one
copied in.

QuentinBisson committed:
Merge branch 'main' into try-multi-arch

jgerber-sbb commented:
It looks like the newest released/tagged version is v0.20.1 which is still using giantswarm/architect@6.12.0
--> no arm64 support yet
Are there plans for a new release?
